python_scripts/ensemble_hyperparameters.py

```diff
@@ -136,48 +131,52 @@
+# one after the other anymore. The important hyperparameters are `max_iter`,
 # `learning_rate`, and `max_depth` or `max_leaf_nodes` (as previously discussed
 # random forest).
 #
-# Let's first discuss the `max_depth` (or `max_leaf_nodes`) parameter. We saw in
-# the section on gradient-boosting that the algorithm fits the error of the
-# previous tree in the ensemble. Thus, fitting fully grown trees would be
+# Let's first discuss the `max_iter`, which similarly to the `n_estimators`
+# hyperparameter in random forest controls the number of trees in the estimator.
+# The difference is that the actual number of trees required by the model is
+# not directly set by the user but depends on the stopping criteria.
+#
+# The depth of the trees is controlled by `max_depth` (or `max_leaf_nodes`). We
+# saw in the section on gradient-boosting that boosting algorithms fit the error
+# of the previous tree in the ensemble. Thus, fitting fully grown trees would be
 # detrimental. Indeed, the first tree of the ensemble would perfectly fit
 # (overfit) the data and thus no subsequent tree would be required, since there
 # would be no residuals. Therefore, the tree used in gradient-boosting should
 # have a low depth, typically between 3 to 8 levels, or few leaves ($2^3=8$ to
-# $2^8=256$). Having very weak learners at each step will help reducing
-# overfitting.
+# $2^8=256$). Having very weak learners at each step helps reducing overfitting.
 #
 # With this consideration in mind, the deeper the trees, the faster the
-# residuals will be corrected and less learners are required. Therefore,
-# `n_estimators` should be increased if `max_depth` is lower.
+# residuals are corrected and then less learners are required. Therefore,
+# `max_iter` should be increased if `max_depth` is lower.
 #
 # Finally, we have overlooked the impact of the `learning_rate` parameter until
 # now. When fitting the residuals, we would like the tree to try to correct all
 # possible errors or only a fraction of them. The learning-rate allows you to
 # control this behaviour. A small learning-rate value would only correct the
 # residuals of very few samples. If a large learning-rate is set (e.g., 1), we
 # would fit the residuals of all samples. So, with a very low learning-rate, we
-# will need more estimators to correct the overall error. However, a too large
-# learning-rate tends to obtain an overfitted ensemble, similar to having a too
-# large tree depth.
+# would need more estimators to correct the overall error. However, a too large
+# learning-rate tends to obtain an overfitted ensemble, similar to having very
+# deep trees.
 
 # %%
 from scipy.stats import loguniform
-from sklearn.ensemble import GradientBoostingRegressor
+from sklearn.ensemble import HistGradientBoostingRegressor
 
 param_distributions = {
-    "n_estimators": [1, 2, 5, 10, 20, 50, 100, 200, 500],
+    "max_iter": [2, 5, 10, 20, 50, 100, 200, 500],
```

Comment:
Maybe we can try a coarser grid but going to a larger maximum to see if the best model can reach this bound or not.

```suggestion
    "max_iter": [3, 10, 30, 100, 300, 1000],
```